core/src/bundle_stage.rs

```diff
@@ -122,6 +122,80 @@ struct AllExecutionResults {
     pub post_balances: (TransactionBalances, TransactionTokenBalances),
 }
 
+struct BundleReservedSpace {
+    current_allocated_cost: u64,
+    initial_allocated_cost: u64,
+}
+
+impl BundleReservedSpace {
+    fn reset_reserved_cost(&mut self, working_bank: &Arc<Bank>) {
+        self.current_allocated_cost = self.initial_allocated_cost;
+
+        // Reserve space from cost tracker
+        QosService::update_or_remove_transaction_costs(
+            vec![TransactionCost::new_with_capacity(1)].iter(),
+            vec![Ok(())].iter(),
+            Some(&vec![CommitTransactionDetails::Committed {
+                compute_units: self.current_allocated_cost,
+            }]),
+            working_bank,
+        );
+        info!(
+            "Slot: {} - Update reserved space to {}. Block Cost {}",
+            working_bank.slot(),
+            self.current_allocated_cost,
+            working_bank
+                .read_cost_tracker()
+                // Using unwrap here is scary, but that's the way it's done in QoS
+                .unwrap()
+                .block_cost()
+        );
+    }
+
+    fn use_reserved_cost(&mut self, bundle_cost: u64, working_bank: &Arc<Bank>) {
+        // If there is enough reserved space left, deduct bundle cost from reserved space before adding
+        if bundle_cost <= self.current_allocated_cost {
+            self.current_allocated_cost -= bundle_cost;
+            let mut dummy_tx_cost = TransactionCost::new_with_capacity(1);
+            dummy_tx_cost.bpf_execution_cost = bundle_cost;
+            QosService::remove_transaction_costs(
```

Comment:
Should this be calling `update_transaction_costs`?

Reply:
that would just require an extra definition of an estimated tx cost, a comparison, then passing the difference into cost_tracker.sub_transaction_execution_cost

remove_transaction_cost results in identical behavior from the cost_tracker perspective